### DISCUSSION TITLE
ignore output of part of a rule

### DISCUSSION THREAD

**Discussion by larsimmisch** [taocpp/PEGTL#51] [Q&A] [answered]
First of all: this is great. I haven't had that much fun writing a grammar for ages.

I want to use PEGTL for something relatively straightforward, parsing Lua tables.

I've got something working based upon your Lua example and I'm happy with the progress I've made in just a few hours.

But as I am getting into attaching actions to the rules (to construct values from the parser run), I see that I will hit a snag with the unary plus/minus operator that I need for numerals. Unfortunately, `- --[[ this is silly --]] 2` is valid Lua and `-2`.

In my rule:
```
	struct decimal_signed : p::seq< p::one< '+', '-' >, seps, decimal > {};
```

I'd like to ignore the `seps`, i.e. I want to match it, but not have it produce any output. What is the best way to achieve this?

**Comment by ColinH** [comment]
First of all: Thanks for the kind words, we are glad that you enjoy using the PEGTL :-)

Regarding your question, there is (currently) no way to obtain only part of the matched input, short of attaching multiple actions to the sub-rules, which should be the easiest way to achieve what you need:

Let the action attached to the`p::one< '+', '-' >`-rule remember the sign in a state, then do everything else in the action attached to `decimal`.

(Presumably you need the sign and the digits together in order to to call a standard string-to-integer function; in that case, since the data might not be contiguous in the input, it will have to be somehow "copied together". Or you can roll your own string-to-integer function spread over the two actions...)

**Comment by d-frey** [comment, marked as answer]
Adding another idea:

    struct decimal_plus : decimal {};
    struct decimal_minus : decimal {};
    struct decimal_plus_full : p::seq< p::one< '+' >, seps, decimal_plus > {};
    struct decimal_minus_full : p::seq< p::one< '-' >, seps, decimal_minus > {};
    struct decimal_signed : p::sor< decimal_plus_full, decimal_minus_full > {};

Then attach your actions to `decimal_plus` and `decimal_minus`. In the latter case, just convert the input to a number and negate that (positive) number after a proper range check.

**Comment by larsimmisch** [comment]
Nice! Thank you!

I think I'll use a variant from @d-frey with actions attached to `decimal_minus` and `decimal_nosign`:

```
struct decimal_nosign : numeral_one< p::digit, p::one< 'e', 'E' > > {};
struct decimal_minus : decimal_nosign {};
struct decimal_plus_full : p::seq< p::one< '+' >, seps, decimal_nosign > {};
struct decimal_minus_full : p::seq< p::one< '-' >, seps, decimal_minus > {};
struct decimal : p::sor< decimal_plus_full, decimal_minus_full, decimal_nosign > {};
```

**Comment by d-frey** [comment]
You're welcome. If similar problems come up more often in the future and the work-arounds get out of hand, feel free to re-open this issue.
